docs/02-basic-tutorial.md

```diff
@@ -0,0 +1,363 @@
+```
+
+Now we need to add some state handling to make it interactive
+
+```livescript
+matches = (query, item) ––>
+  item.index-of query > 0
+
+module.exports = reflex.create-component do
+  display-name: ‘listing’
+
+  get-initial-state: ->
+    query: ‘’
+
+  render: ->
+    d.div do
+      d.h1 “A list of useful things“
+      d.form do
+        d.input do
+          type: ‘search’
+          placeholder: ‘Search things’
+          value: @state.query
+          on-change: ~> @set-state query: it.target.value
+      d.ul do
+        things
+        |> filter matches @state.query
+        |> map ->
+          d.li it
+```
+
+Notice the use of [currying](http://livescript.net/#functions-currying) in the definition of the `matches` function.
+
+React doesn’t go much further than that. In real-world applications however, state rarely spans just a single component. The existence of the Flux architecture is good evidence of the fact.
+
+## Centralised state instead of Flux
+
+Reflex takes a different approach to state, which is very similar to the [Om framework](https://github.com/omcljs/om) for ClojureScript. In Reflex, all shared UI state is kept in a single place, the `app-state` - application state.
+
+The application state is a “cursor” - a focused view of a part of a larger data structure that can be mutated in controlled fashion. There is a larger discussion of the application structure in the [Reflex Architecture](04-reflex-architecture.md) section.
+
+Let’s add a list of recent searches into our little demo. Since it will be another listing, we should keep our code DRY and extract the list rendering into a separate component.
+
+```livescript
+# list.ls
+
+require! <[ reflex ]>
+d = reflex.DOM
+
+module.exports = reflex.create-component do
+  display-name: ‘list’
+  render: ->
+    d.ul
+      @this.props.item |> map ->
+        d.li it
+```
+
+Then we can use it in our listing route component
+
+```livescript
+  render: ->
+    d.div do
+      d.h1 “A list of useful things“
+      d.form do
+        d.input do
+          type: ‘search’
+          placeholder: ‘Search things’
+          value: @state.query
+          on-change: ~> @set-state query: it.target.value
+      list do
+        items: (things |> filter matches @state.query)
+```
+
+Actually, the filtering functionality seems very common, lets include that in the component too. The filtering logic itself is very use-case specific though, so it should be external.
+
+This is now our listing component
+
+```livescript
+  render: ->
+    d.div do
+      d.h1 “A list of useful things“
+      list do
+        query: @state.query
+        items: (things |> filter matches @state.query)
+```
+
+and the filterable list
+
+```livescript
+  render: ->
+    d.div do
+      d.form do
+        d.input do
+          type: ‘search’
+          placeholder: ‘Search things’
+          value: @props.query
+          on-change: ~> # now what?
+      ul do
+        @props.things |> map -> li it
+```
```

Comment:
these two code excerpts are sort of belongs to the same `listing.ls` component at the moment of arriving to this point.